// @tanstack/router-vite-plugin/dist/cjs/eliminateUnreferencedIdentifiers.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const t = require("@babel/types");
function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: "Module" } });
  if (e) {
    for (const k in e) {
      if (k !== "default") {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}
const t__namespace = /* @__PURE__ */ _interopNamespaceDefault(t);
const eliminateUnreferencedIdentifiers = (programPath, refs) => {
  let referencesRemovedInThisPass;
  const shouldBeRemoved = (ident) => {
    if (isIdentifierReferenced(ident))
      return false;
    if (!refs)
      return true;
    return refs.has(ident);
  };
  const sweepFunction = (path) => {
    const identifier = getIdentifier(path);
    if ((identifier == null ? void 0 : identifier.node) && shouldBeRemoved(identifier)) {
      ++referencesRemovedInThisPass;
      if (t__namespace.isAssignmentExpression(path.parentPath.node) || t__namespace.isVariableDeclarator(path.parentPath.node)) {
        path.parentPath.remove();
      } else {
        path.remove();
      }
    }
  };
  const sweepImport = (path) => {
    const local = path.get("local");
    if (shouldBeRemoved(local)) {
      ++referencesRemovedInThisPass;
      path.remove();
      if (path.parent.specifiers.length === 0) {
        path.parentPath.remove();
      }
    }
  };
  do {
    referencesRemovedInThisPass = 0;
    programPath.scope.crawl();
    programPath.traverse({
      VariableDeclarator(path) {
        if (path.node.id.type === "Identifier") {
          const local = path.get("id");
          if (shouldBeRemoved(local)) {
            ++referencesRemovedInThisPass;
            path.remove();
          }
        } else if (path.node.id.type === "ObjectPattern") {
          const pattern = path.get("id");
          const beforeCount = referencesRemovedInThisPass;
          const properties = pattern.get("properties");
          properties.forEach((property) => {
            const local = property.get(
              property.node.type === "ObjectProperty" ? "value" : (
                // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
                property.node.type === "RestElement" ? "argument" : function() {
                  throw new Error("invariant");
                }()
              )
            );
            if (shouldBeRemoved(local)) {
              ++referencesRemovedInThisPass;
              property.remove();
            }
          });
          if (beforeCount !== referencesRemovedInThisPass && pattern.get("properties").length < 1) {
            path.remove();
          }
        } else if (path.node.id.type === "ArrayPattern") {
          const pattern = path.get("id");
          const beforeCount = referencesRemovedInThisPass;
          const elements = pattern.get("elements");
          elements.forEach((e) => {
            var _a, _b;
            let local;
            if (((_a = e.node) == null ? void 0 : _a.type) === "Identifier") {
              local = e;
            } else if (((_b = e.node) == null ? void 0 : _b.type) === "RestElement") {
              local = e.get("argument");
            } else {
              return;
            }
            if (shouldBeRemoved(local)) {
              ++referencesRemovedInThisPass;
              e.remove();
            }
          });
          if (beforeCount !== referencesRemovedInThisPass && pattern.get("elements").length < 1) {
            path.remove();
          }
        }
      },
      FunctionDeclaration: sweepFunction,
      FunctionExpression: sweepFunction,
      ArrowFunctionExpression: sweepFunction,
      ImportSpecifier: sweepImport,
      ImportDefaultSpecifier: sweepImport,
      ImportNamespaceSpecifier: sweepImport
    });
  } while (referencesRemovedInThisPass);
};
function getIdentifier(path) {
  const parentPath = path.parentPath;
  if (parentPath.type === "VariableDeclarator") {
    const variablePath = parentPath;
    const name = variablePath.get("id");
    return name.node.type === "Identifier" ? name : null;
  }
  if (parentPath.type === "AssignmentExpression") {
    const variablePath = parentPath;
    const name = variablePath.get("left");
    return name.node.type === "Identifier" ? name : null;
  }
  if (path.node.type === "ArrowFunctionExpression") {
    return null;
  }
  return path.node.id && path.node.id.type === "Identifier" ? path.get("id") : null;
}
function isIdentifierReferenced(ident) {
  const binding = ident.scope.getBinding(ident.node.name);
  if (binding == null ? void 0 : binding.referenced) {
    if (binding.path.type === "FunctionDeclaration") {
      return !binding.constantViolations.concat(binding.referencePaths).every((ref) => ref.findParent((parent) => parent === binding.path));
    }
    return true;
  }
  return false;
}
exports.eliminateUnreferencedIdentifiers = eliminateUnreferencedIdentifiers;
